sycl/include/CL/sycl/detail/device_impl.hpp

```diff
@@ -21,44 +20,141 @@ namespace sycl {
 class platform;
 
 namespace detail {
-// TODO: 4.6.4 Partitioning into multiple SYCL devices
-// TODO: 4.6.4.2 Device information descriptors
+
+// Forward declaration
+class platform_impl;
+class platform_impl_pi;
+
 // TODO: Make code thread-safe
 class device_impl {
 public:
-  virtual ~device_impl() = default;
-
-  virtual cl_device_id get() const = 0;
-
-  // Returns underlying native device object (if any) w/o reference count
-  // modification. Caller must ensure the returned object lives on stack only.
-  // It can also be safely passed to the underlying native runtime API.
-  // Warning. Returned reference will be invalid if device_impl was destroyed.
-  //
-  virtual RT::PiDevice &getHandleRef() = 0;
-  virtual const RT::PiDevice &getHandleRef() const = 0;
-
-  virtual bool is_host() const = 0;
-
-  virtual bool is_cpu() const = 0;
-
-  virtual bool is_gpu() const = 0;
-
-  virtual bool is_accelerator() const = 0;
-
-  virtual platform get_platform() const = 0;
+  /// Constructs a SYCL device instance as a host device.
+  device_impl();
+  /// Constructs a SYCL device instance using the provided
+  /// PI device instance.
+  explicit device_impl(RT::PiDevice Device);
+
+  ~device_impl();
+
+  /// Get instance of OpenCL device
+  ///
+  /// @return a valid cl_device_id instance in accordance with the requirements
+  /// described in 4.3.1.
+  cl_device_id get() const;
+
+  /// Get reference to PI device
+  ///
+  /// For host device an exception is thrown
+  ///
+  /// @return non-constant reference to PI device
+  RT::PiDevice &getHandleRef() {
+    if (MIsHostDevice)
+      throw invalid_object_error("This instance of device is a host instance");
+
+    return MDevice;
+  }
+
+  /// Get constant reference to PI device
+  ///
+  /// For host device an exception is thrown
+  ///
+  /// @return constant reference to PI device
+  const RT::PiDevice &getHandleRef() const {
+    if (MIsHostDevice)
+      throw invalid_object_error("This instance of device is a host instance");
+
+    return MDevice;
+  }
+
+  /// Check if device is a host device
```

Comment:
```suggestion
  /// Check if SYCL device is a host device
```
Same for the related functions below